### PR TITLE
SOR E2E tests

### PR DESCRIPTION
**Pull request by timjrobinson**
This adds some E2E tests for SOR that does the following:

- Connects to the /sor endpoint
- Queries a swap of tokens
- Takes the response and turns it into on-chain transactions
- Performs these transactions in a Hardhat fork and reports if they succeeded or not. 

The aim is to create canary tests that can inform us if the swaps returned by the SOR endpoint are invalid (as Cowswap reported once before) and so that we can test new versions of or updates to the SDK / SOR. 

How to use:

- Set `RPC_URL` in your .env file to Infura/Alchemy/Local Node. Or set `INFURA_PROJECT_ID`
- Set `ENDPOINT_URL` to your dev API, defaults to `https://api.balancer.fi`

```
npm run node
npm run test:e2e
```

TODO:
- [x] Add info to README
- [x] Add tests for many more swap paths
- [x] Clean up and remove redundancies in code

### CODE REVIEW

**Comment by alter-eggo** [comment]
tests work ok for me

**Comment by timjrobinson** [comment]
I've added a bunch of tests for common paths and using the bb-a-USD pool and cleaned up the code. It's ready for review now.

File: test/e2e/sor.spec.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        buyToken: BAL.address,
+        orderKind: 'sell',
+        amount: parseFixed('100', 18).toString(),
+        gasPrice: BigNumber.from('0x174876e800').toString(),
+      };
+      await setTokenBalance(signer, DAI, sorRequest.amount)
+      const balances = await getBalances(signer, [BAL]);
+      await testSorRequest(signer, Network.MAINNET, sorRequest);
+      const newBalances = await getBalances(signer, [BAL]);
+      expect(BigNumber.from(newBalances.BAL).gt(balances.BAL));
```
Comment (garethfuller) [inline test/e2e/sor.spec.ts:63]: Should we not be testing that the amount received is a specific amount? The problem we had with routing through boosted pools (described in the asana ticket) was that the SOR was returning incorrect values. Will these tests failif that were to happen again?

**Review by garethfuller — APPROVED** [review]
Looks good. I managed to run locally.

File: test/e2e/sor.spec.ts (hunk 1 above)
Comment (timjrobinson) [reply test/e2e/sor.spec.ts:63]: It's really checking if the batch swap fails. Which should happen if the numbers are bad. Though I could add a sanity check test that makes sure it fails when the numbers are off.

**Review by alter-eggo — APPROVED** [review]

File: test/e2e/sor.spec.ts (hunk 1 above)
Comment (timjrobinson) [reply test/e2e/sor.spec.ts:78]: I'd added a test that purposely fails. It also revealed that my limit settings were wrong which is good.

**Pull request merged**